### PR TITLE
SHINY: migrate from Chandra.cmd_states / Ska.ParseCM to kadi

### PR DESCRIPTION
**Pull request by taldcroft**
## Description

This includes changes necessary to support migrating from Chandra.cmd_states / Ska.ParseCM to kadi in conjunction with
https://github.com/acisops/acis_thermal_check/pull/30.

It also has a somewhat separate fix to allow running with an OFLS dir that is not the hardwired `/data/acis`. If desired this could be pulled out in to a separate PR.

## Testing

- [N/A] Passes unit tests (no unit tests)
- [x] Functional testing via https://github.com/acisops/acis_thermal_check/pull/30

### CODE REVIEW

File: backstop_history/BackstopHistory.py [hunk 1]
```diff
@@ -298,68 +299,80 @@ def print_backstop_lists(self):
...
             ofls_cont_file = open(ofls_cont_fn, 'r')
-            # Read the first line...the pat to the continuity load
-            continuity_load_path = ofls_cont_file.readline()[:-1]
-    
+
+            # Read the first line...the path to the continuity load. The
+            # continuity path in the file is hardwired to a /data/acis path,
+            # independent of user-specified `oflsdir` (e.g.
+            # /data/acis/LoadReviews/2018/MAY2118/ofls), so fix that by putting
+            # the last 3 parts (2018/MAY2118/ofls) onto the oflsdir root.
```
Comment (jzuhone) [inline backstop_history/BackstopHistory.py:352]: The `backstop_history` machinery doesn't work unless the path is rooted in `/data/acis`. This is because the link from the latest load directory (say) `oflsa` to `ofls` is linked to the absolute path. This is also true for other paths in this package.

Comment (taldcroft) [reply backstop_history/BackstopHistory.py:352]: It worked OK for me and I was able to use the backstop_history machinery on my Mac with the following (somewhat contorted) setup:

- I use sshfs to make a local mount `~/kady` on my Mac laptop that points to `kady:/home/aldcroft`
- On kady:
  - `/home/aldcroft/ska -> /proj/sot/ska`
  - `/proj/sot/ska/data/acis -> /data/acis`

With that it seems to work fine:
```
ls ~/kady/ska/data/acis/LoadReviews/2018/MAY2118/ofls 
ACIS-Continuity.txt        ACIS-Perigee.txt           LRHS.output                fp_sensitive.txt           out_fptemp/
ACIS-DITHHIST.dat          ACIS-STORED-HIST.txt       MAY2118A_CTI_report.txt*   mmMAY2118A.sum*            out_psmc/
ACIS-FPHIST.dat            ACIS-TLMHIST.dat           MAY2118A_backstop.tar.gz   msMAY2118A.sum*            output/
ACIS-GRATHIST.dat          ACIS-TSCHIST.dat           MAY2118A_lr.html           out_bep_pcb/               perigee_times.pro
ACIS-History.txt           ACIS-Tables.txt            MAY2118_angles.txt         out_dea/                   saa_check.txt
ACIS-LoadReview.txt        CMAY2118_02.er*            MAY2118_angles_idl.ps      out_dpa/                   vehicle/
ACIS-LoadReview_HandS.txt  CR141_0311.backstop        cr_test.backstop           out_fep1_actel/            violations.txt
ACIS-OBSHIST.dat           DO18113_CRM_Pad.txt*       dahtbon_history.rdb        out_fep1_mong/             vr_test.backstop
```

And finally in this context the following gave an expected result:
```
dpa_check \
   --outdir=headout \
   --oflsdir=/Users/aldcroft/kady/ska/data/acis/LoadReviews/2018/MAY2818/oflsa \
   --run-start=2018:142
```

Is there any reason that the `ofls` links are absolute?

Comment (Gregg140) [reply backstop_history/BackstopHistory.py:352]: > Is there any reason that the ofls links are absolute?

Yes.

The ACIS Ops Load Review program (lr) creates directories for all of its products with the root of the path being /data/acis/LoadReviews/ followed by the year (e.g. 2020) and Weekly Load names (e.g. JUL1320) followed by ofls names (e.g. oflsa).  It has been this way since 2005.  

Among these products are the load continuity information and interrupt times(if any), captured and recorded in each pertinent load directory by lr.  This  is used by BackstopHistory. At this point in the mission, the chances of the load review directory structure changing are so remote as to be almost non-existent: some catastrophic event would have to occur to force a change.  

There are no plans nor expectation of changing BackstopHistory to acquire the continuity and interrupt information in any other way.  It has worked for several years; is internal to ACIS Ops; and therefore under ACIS Ops control and is changed only if a bug is found.

The only changes  to BackstopHistory presently contemplated are ones generated by changes to  externally controlled software which:

Eliminates Ska.ParseCM
Deprecation of the Commanded States Database and the move to KADI for that information.

Comment (jzuhone) [reply backstop_history/BackstopHistory.py:352]: @Gregg140 unless I misread what you were saying, @taldcroft's question was slightly different. It seems like his question is about linking between `ofls` and `oflsa`, etc., if it's necessary to use the full, absolute path starting with `/data/acis/...` or if links using the relative path would suffice. If we did the latter, one could re-create the structure of `LoadReviews` somewhere else for testing without having to place it under `/data/acis`. Given testing strategies going forward, this could be potentially helpful. 

On the other hand, I realize changing all of the links which presently exist is a huge pain and potentially prone to causing errors.

**Comment by taldcroft** [comment]
I certainly could open a separate PR to address the changes not related to the Ska.ParseCM deprecation, but this is honestly not the most efficient use of resources, and follows common ACIS ops practice of getting into a script and making various improvements along the way. In particular, I consider it to be bug that the `--oflsdir=...` flag does not actually work unless that directory is within `/data/acis`.  I needed that functionality in order to fully test my changes related to Ska.ParseCm locally on my Mac. It seems like a useful bug fix / enhancement to me with really no downside.

**Comment by jzuhone** [comment]
@taldcroft I'm fine with leaving this PR as-is--the team can discuss the separate issue of the symlinks offline. We can focus on making sure the tests pass (there is the lingering issue I pointed out [here](https://github.com/acisops/acis_thermal_check/pull/30#issuecomment-660192373)).

**Comment by taldcroft** [comment]
@jzuhone - agreed that the symlinks issue is separate and doesn't impact this at all. To close the loop here, I believe the lingering issue you mentioned is fixed now.

**Comment by jzuhone** [comment]
Hi @taldcroft--@Gregg140 and I discussed this and we'd actually like to punt on the code that changes the root directory for now (but let's leave the `pathlib` directory handling in, I need to start using it myself). We determined that this would lead to a lot of downstream testing of not only this package but also the perl-based LR software and we'd like to handle this all at once at some point in the future rather than just putting this one piece in now (because I am sure we'd find lots of corner cases, etc.). I'd prefer not to hold up merging this PR longer than we have to given we want to promote shiny to flight by August 25th. 

For testing on non-HEAD platforms, we can create (or sshfs-mount) `/data/acis` on Linux, and what I've done for testing this package on my Mac is rsync the relevant directories under `/data/acis` to a directory (say) `${HOME}/head_data/acis` and then use the `/etc/synthetic.conf` file to map `${HOME}/head_data/acis` to `/data/acis`:

https://derflounder.wordpress.com/2020/01/18/creating-root-level-directories-and-symbolic-links-on-macos-catalina/

You could probably do the same thing with a sshfs mount. 

Because we can't create a `/data/acis` on Windows we'd have to punt on Windows testing for `backstop_history` for now, but I also don't see us using it anytime soon either. 

I plan on getting back into testing this today so we can green-light this soon. 

How's all that sound?

**Comment by taldcroft** [comment]
@jzuhone - sure.  I'm not thrilled but it's a way forward and I'm glad you are working on this.

I do think it a good thing to be able to fully test the software on machines without having to re-configure each machine itself with hacky tricks. I did once try this `/etc/synthetic.conf` thing a while back and got no joy. I'm sure I did something wrong but I'm not planning to go that route.  Instead with my 2-line patch, as long as you have the right data in any Ska-based location, tests can pass. I guess I'm not cognizant of the corner cases that might make things go wrong.

**Comment by taldcroft** [comment]
@jzuhone - I added a commit 7ca81b7 that insists on the current production behavior that `oflsdir` must be in the `/data/acis` root unless the environment variable `ALLOW_NONSTANDARD_OFLSDIR` is set. So the code is transparently maintaining the current behavior (modulo using `Path`), but allows development and testing on a standalone platform without needing to reconfigure the host machine.  Is that OK?  

I tested and confirmed:
- Fails as expected with a non-standard path and no env var set (on Mac and HEAD)
- Runs as expected with a non-standard path and the env var set (on Mac)
- Runs as expected with a standard /data/acis path and no env var set (on HEAD)
